### PR TITLE
SprkTextInput aria-describedby

### LINKED ISSUE #3526: SprkInputElement should only have aria-describedby if the error container exists
## Describe The Bug
`SprkInputElement` is hard-coded to have aria-describedby even if the id provided does not match an existing error container. This causes an a11y violation.

## To Reproduce
1. Go to https://react.sparkdesignsystem.com/iframe.html?id=components-input-text--text-input
2. Open Chrome Dev Tools > Audits
3. Click "Generate Report"

Lighthouse report:
image

Source code:
image

## Expected Behavior
- SprkInputElement should only render with aria-describedby if the id is valid.
- New unit tests that test for this functionality

### PR DESCRIPTION
**Pull request by bobwise**
~~Modifies SprkTextInput to only include `aria-describedby` on the `<input>` if it has an error container.~~

New scope: Refactor `SprkTextInput` and `SprkInputElement` to support these use cases:
1. User can set their own value(s) for `aria-describedby` on the input.
2. If the input is invalid, the component will merge the error container ID with any values provided by the user.
3. `aria-describedby` will only include the error container ID if the value is invalid.
4. `aria-describedby` will only be calculated if `errorContainerId` or `ariaDescribedBy` have been updated.

### Associated Issue
Fixes #3526 (also https://github.com/sparkdesignsystem/spark-design-system/issues/3585)

### Code
 - [x] Build Component in React
 - [x] Unit Testing in React with `npm run test` in `react/` (100% coverage, 100% passing)

### Browser Testing (current version and 1 prior)
  - [x] Google Chrome
  - [x] Google Chrome (Mobile)
  - [x] Mozilla Firefox
  - [x] Mozilla Firefox (Mobile)
  - [x] Microsoft Internet Explorer 11 (only this specific version of IE)
  - [x] Microsoft Edge
  - [x] Apple Safari
  - [x] Apple Safari (Mobile)
  
### Accessibility Testing
  - [x] Axe browser extension
  - [x] Jaws Inspect
  - [x] VoiceOver (iOS) or Talkback (Android)

### Deploy Preview Reviewed and Approved
 - [x] Accessibility Team

### CODE REVIEW

**Review by afebbraro — CHANGES_REQUESTED** [review]
aria-described by could be used for more use cases besides the field being invalid. If we only allow it if there’s an error container then we could be restricting use. What do you think?

**Review by guern1kn — APPROVED** [review]
I think that @afebbraro has a good point, but I think it's fine the way it currently is. We are currently setting the `errorContainerId` internally and passing it to the `<SprkInputElement>`. Both the `<SprkTextInput>` and `<SprkInputElement>` are in the process of being replaced. I don't know that it makes sense to spend time re-writing these components to allow for additional functionality at this point. Amber's point would be something to keep in mind for the React Input API upgrades though.

**Comment by bobwise** [comment]
Yes @afebbraro is correct. `aria-describedby` can actually have multiple values passed into it at the same time, which Spark should support. Spark should be putting the helper text container id in the `aria-describedby` as well as allowing users to add their own. This is one of the items that came up in our accessibility audit (HTML framework). https://github.com/sparkdesignsystem/Backlog/issues/1332

File: react/src/base/inputs/SprkTextInput/SprkTextInput.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ class SprkTextInput extends Component {
               forwardedRef={forwardedRef}
               type={type}
               formatter={formatter}
-              errorContainerId={errorContainerId}
+              errorContainerId={valid ? undefined : errorContainerId}
```
Comment (RVMendoza) [inline react/src/base/inputs/SprkTextInput/SprkTextInput.js:70]: Nice! This is a good way to make sure screen readers aren't confused if the error doesn't exist.

**Review by RVMendoza — APPROVED** [review]

**Comment by afebbraro** [comment]
@bobwise  If we want to fix the bug and not introduce new a11y errors, it would be best to update this component to accept an ariaDescribedBy prop that if it has a value would pass down to the input element component and additionally, if the input is also invalid it would join errorContainerId and the ariaDescribedBy into a string that it would pass down. That away, users can supply their own ariaDescribedBy and we would also linking the error container correctly.

**Review by guern1kn — APPROVED** [review]

**Review by RVMendoza — APPROVED** [review]

File: react/src/base/inputs/components/SprkInputElement/SprkInputElement.js [hunk 2]
```diff
@@ -51,6 +52,18 @@ class SprkInputElement extends Component {
       }
     };
 
+    const getAriaDescribedBy = () => {
```
Comment (RVMendoza) [inline react/src/base/inputs/components/SprkInputElement/SprkInputElement.js:55]: nice! this is definitely something we should be incorporating to our other components that use aria-describedby

File: react/src/base/inputs/components/SprkInputElement/SprkInputElement.js [hunk 3]
```diff
@@ -73,9 +86,9 @@ class SprkInputElement extends Component {
         data-id={idString}
         data-analytics={analyticsString}
         aria-invalid={!valid}
-        aria-describedby={errorContainerId}
+        aria-describedby={getAriaDescribedBy()}
```
Comment (afebbraro) [inline react/src/base/inputs/components/SprkInputElement/SprkInputElement.js:89]: Thanks for adding this. I did notice a potential issue. This seems to be running every time the component renders for me, so even though the errorcontainer /id hasnt changed this gets triggered. Every time I type a character in the input, it runs. Which seems to drag down performance. Maybe I am doing something weird? But thats what I am seeing.  I think something to try here would be to bind this to a method that runs on one of the lifecycle hooks, componentDidMount() ?

**Review by afebbraro — APPROVED** [review]

**Pull request merged**